src/PotatoTcp/HandlerStrategies/MemoizingBaseHandlerStrategy.cs

```diff
@@ -0,0 +1,118 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace PotatoTcp.HandlerStrategies
+{
+    public class MemoizingBaseHandlerStrategy : IHandlerStrategy
+    {
+        private readonly ConcurrentDictionary<Type, Type> _handlerTypes = new ConcurrentDictionary<Type, Type>();
+        private readonly ConcurrentDictionary<Type, List<IMessageHandler>> _handlers = new ConcurrentDictionary<Type, List<IMessageHandler>>();
+
+        public void AddHandler<T>(Action<Guid, T> handler)
+        {
+            var handlerType = typeof(T);
+            var messageHandler = new MessageHandler<T>
+            {
+                HandlerType = handlerType,
+                HandlerAction = handler
+            };
+
+            _handlers.AddOrUpdate(
+                handlerType,
+                new List<IMessageHandler> {messageHandler},
+                (type, handlers) =>
+                {
+                    handlers.Add(messageHandler);
+                    return handlers;
+                });
+        }
+
+        public void AddHandler(IMessageHandler handler)
+        {
+            _handlers.AddOrUpdate(
+                handler.HandlerType,
+                new List<IMessageHandler> {handler},
+                (type, handlers) =>
+                {
+                    handlers.Add(handler);
+                    return handlers;
+                });
+        }
+
+        public bool InvokeHandler(object message)
+        {
+            var messageType = message.GetType();
+
+            if (TryInvokeHandler(messageType, message)) return true;
+
+            if (_handlerTypes.TryGetValue(messageType, out Type baseHandlerType))
+            {
+                if (TryInvokeHandler(baseHandlerType, message)) return true;
+
+                var firstBaseType = GetBaseTypes(messageType).FirstOrDefault(x => _handlers.TryGetValue(x, out _));
+                _handlerTypes.AddOrUpdate(messageType, _ => firstBaseType, (_, __) => firstBaseType);
+                return TryInvokeHandler(firstBaseType, message);
+            }
+            return false;
```

Comment:
I think we're missing an else statement here.
- If there are handlers for this type, execute them.
- If there is a base type mapping for this type, execute those handlers
- **else get the base type mapping and recursively call this method**?

Reply:
I think some of that other code moved into the above if when it wasn't suppose to...